x/concentrated-liquidity/types/pool_hooks.go

```diff
@@ -0,0 +1,152 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+)
+
+// Action prefixes for pool actions
+const (
+	CreatePosition     = "createPosition"
+	AddToPosition      = "addToPosition"
+	WithdrawPosition   = "withdrawPosition"
+	SwapExactAmountIn  = "swapExactAmountIn"
+	SwapExactAmountOut = "swapExactAmountOut"
+)
+
+// --- Sudo Message Wrappers ---
+
+type BeforeCreatePositionSudoMsg struct {
+	BeforeCreatePosition BeforeCreatePositionMsg `json:"before_create_position"`
+}
+
+type AfterCreatePositionSudoMsg struct {
+	AfterCreatePosition AfterCreatePositionMsg `json:"after_create_position"`
+}
+
+type BeforeAddToPositionSudoMsg struct {
+	BeforeAddToPosition BeforeAddToPositionMsg `json:"before_add_to_position"`
+}
+
+type AfterAddToPositionSudoMsg struct {
+	AfterAddToPosition AfterAddToPositionMsg `json:"after_add_to_position"`
+}
+
+type BeforeWithdrawPositionSudoMsg struct {
+	BeforeWithdrawPosition BeforeWithdrawPositionMsg `json:"before_withdraw_position"`
+}
+
+type AfterWithdrawPositionSudoMsg struct {
+	AfterWithdrawPosition AfterWithdrawPositionMsg `json:"after_withdraw_position"`
+}
+
+type BeforeSwapExactAmountInSudoMsg struct {
+	BeforeSwapExactAmountIn BeforeSwapExactAmountInMsg `json:"before_swap_exact_amount_in"`
+}
+
+type AfterSwapExactAmountInSudoMsg struct {
+	AfterSwapExactAmountIn AfterSwapExactAmountInMsg `json:"after_swap_exact_amount_in"`
+}
+
+type BeforeSwapExactAmountOutSudoMsg struct {
+	BeforeSwapExactAmountOut BeforeSwapExactAmountOutMsg `json:"before_swap_exact_amount_out"`
+}
+
+type AfterSwapExactAmountOutSudoMsg struct {
+	AfterSwapExactAmountOut AfterSwapExactAmountOutMsg `json:"after_swap_exact_amount_out"`
+}
+
+// --- Message structs ---
```

Comment:
Note: abstracting fields out further for these messages was difficult without adding further complexity/constraints on the contract spec side, so the current implementation has duplicate fields for the before and after hooks for each message

Open to suggestions on ways to improve this if anyone feels strongly about it

Reply:
Yeah this was what I was referring to in my note but after trying to implement test CW contracts it ended up complicating things in ways that would be unintuitive to developers (i.e. some fields are behind abstractions while others aren't)

Also, when we abstract in this way we lose information about variable names so we can't be too general. As a result, attempts at this approach have lead to more lines of code and imo less readable types